src/tuples.jl

```diff
@@ -0,0 +1,86 @@
+
+@inline nstatic(::Val{N}) where {N} = ntuple(StaticInt, Val(N))
+
+invariant_permutation(::Any, ::Any) = False()
+function invariant_permutation(x::T, y::T) where {N,T<:Tuple{Vararg{StaticInt,N}}}
+    if x === nstatic(Val(N))
+        return True()
+    else
+        return False()
+    end
+end
+
+permute(x::Tuple, perm::Val) = permute(x, static(perm))
+permute(x::Tuple{Vararg{Any}}, perm::Tuple{Vararg{StaticInt}}) = eachop(getindex, perm, x)
+function permute(x::Tuple{Vararg{Any,K}}, perm::Tuple{Vararg{StaticInt,K}}) where {K}
+    if invariant_permutation(perm, perm) === False()
+        return eachop(getindex, perm, x)
+    else
+        return x
+    end
+end
+
+"""
+    eachop(op, args...; iterator::Tuple{Vararg{StaticInt}}) -> Tuple
+
+Produces a tuple of `(op(args..., iterator[1]), op(args..., iterator[2]),...)`.
+"""
+eachop(op, itr, args...) = _eachop(op, itr, args)
+@generated function _eachop(op, ::I, args::A) where {A,I}
```

Comment:
If we wanted to avoid generated functions, we could `eachop` with something like:
```julia
@inline function eachop(op::F, itr::Tuple{I1,I2,Vararg}, args::Vararg{Any,K}) where {F,I1,I2,K}
    (op(first(itr), args...), eachop(op, Base.tail(itr), args...)...)
end
@inline function eachop(op::F, itr::Tuple{I}, args::Vararg{Any,K}) where {F,I,K}
    (op(first(itr), args...), )
end
```

Reply:
That would definitely be preferable. I'll do some checks to make sure splatting doesn't ruin inference.

Reply:
Checked it on ArrayInterface.jl and inference still works. Thanks!

Reply:
I added the `::F` and `::Vararg{Any,K}` to try and make sure it specializes.

Feel free to merge and issue a release when ready.